tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ConstSimplifierBase.scala

```diff
@@ -29,49 +29,170 @@ abstract class ConstSimplifierBase {
+      if (leftEx == rightEx) ValEx(TlaInt(0))(intTag) else ex // this actually happens
 
     case OperEx(TlaArithOper.mult, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
-      if (left == 0 || right == 0) {
-        ValEx(TlaInt(BigInt(0)))(intTag)
+      ValEx(TlaInt(left * right))(intTag)
+
+    // 0 * x = 0
+    // 1 * x = x
+    case OperEx(TlaArithOper.mult, ValEx(TlaInt(left)), rightEx) =>
+      if (left == 0) {
+        ValEx(TlaInt(0))(intTag)
       } else if (left == 1) {
-        ValEx(TlaInt(right))(intTag)
+        rightEx
+      } else {
+        OperEx(TlaArithOper.mult, ValEx(TlaInt(left))(intTag), rightEx)(intTag)
+      }
+
+    // x * 0 = 0
+    // x * 1 = x
+    case OperEx(TlaArithOper.mult, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        ValEx(TlaInt(0))(intTag)
+      } else if (right == 1) {
+        leftEx
+      } else {
+        OperEx(TlaArithOper.mult, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    case ex @ OperEx(TlaArithOper.div, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Division by zero at ${ex.toString}")
+      } else {
+        ValEx(TlaInt(left / right))(intTag)
+      }
+
+    // 0 / x = 0
+    case OperEx(TlaArithOper.div, ValEx(TlaInt(left)), rightEx) =>
+      if (left == 0) {
+        ValEx(TlaInt(0))(intTag)
+      } else {
+        OperEx(TlaArithOper.div, ValEx(TlaInt(left))(intTag), rightEx)(intTag)
+      }
+
+    // x / 1 = x
+    case ex @ OperEx(TlaArithOper.div, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Division by zero at ${ex.toString}")
+      } else if (right == 1) {
+        leftEx
+      } else {
+        OperEx(TlaArithOper.div, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    // x / x = 1
+    case ex @ OperEx(TlaArithOper.div, leftEx, rightEx) =>
+      if (leftEx == rightEx) {
+        ValEx(TlaInt(1))(intTag)
+      } else {
+        OperEx(TlaArithOper.div, leftEx, rightEx)(intTag)
+      }
+
+    case ex @ OperEx(TlaArithOper.mod, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Mod by zero at ${ex.toString}")
+      } else {
+        ValEx(TlaInt(left % right))(intTag)
+      }
+
+    // x % 1 = 0
+    case ex @ OperEx(TlaArithOper.mod, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Mod by zero at ${ex.toString}")
       } else if (right == 1) {
-        ValEx(TlaInt(left))(intTag)
+        ValEx(TlaInt(0))(intTag)
+      } else {
+        OperEx(TlaArithOper.mod, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    // x % x = 0
+    case OperEx(TlaArithOper.mod, leftEx, rightEx) =>
+      if (leftEx == rightEx) {
+        ValEx(TlaInt(0))(intTag)
       } else {
-        ValEx(TlaInt(left * right))(intTag)
+        OperEx(TlaArithOper.mod, leftEx, rightEx)(intTag)
```

Comment:
Perhaps another case where biding the patter to `ex @ ...` then returning the `ex` rather than having to reconstruct the expression would simplify things a bit?